illumos-utils/src/contract.rs

```diff
@@ -0,0 +1,387 @@
+        }
+    }
+}
+
+// This thread watches for critical events coming from all process
+// contracts held by sled-agent, and reaps (abandons) contracts which
+// become empty. Process contracts are used in conjunction with
+// zone_enter() in order to run commands within non-global zones, and
+// the contracts used for this come from templates that define becoming
+// empty as a critical event.
+pub fn process_contract_reaper(log: Logger) {
+    let watcher = Watcher::new(ContractType::Process);
+    loop {
+        let event = watcher.watch(&log);
+        if event.typ != CT_PR_EV_EMPTY {
+            continue;
+        }
+
+        let ctl = match Control::new(ContractType::Process, event.ctid) {
+            Ok(c) => c,
+            Err(e) => {
+                error!(&log, "Failed to open contract control: {e:?}");
+                continue;
+            }
+        };
+
+        if let Err(e) = ctl.abandon() {
+            error!(log, "{e:?}");
+        } else {
+            debug!(&log, "Abandoned contract {}", event.ctid)
+        }
+    }
+}
+
+// A Rust wrapper around the process contract template.
+#[derive(Debug)]
+pub struct Template {
+    fd: c_int,
+}
+
+impl Drop for Template {
+    fn drop(&mut self) {
+        self.clear();
+        // Ignore any error, since printing may interfere with `slog`'s
+        // structured output.
+        unsafe { libc::close(self.fd) };
+    }
+}
+
+// This follows the symlink from /dev/tfpkt0 to the full path in the /devices
+// tree.  If found, it returns the path as both a string (for error reporting)
+// and as a vector of bytes (for passing to subsequent system calls).  This
+// redundancy comes from illumos/x86_64 expecting characters to be signed,
+// while Rust expects them to be unsigned.
+fn get_tfpkt_device_path() -> Option<(String, Vec<i8>)> {
+    let dev_path = CString::new("/dev/tfpkt0").unwrap();
+    let mut path_buf = [0i8; 1024];
+    let sz = unsafe {
+        libc::readlink(dev_path.as_ptr(), path_buf.as_mut_ptr(), 1024)
+    };
+    if sz < 0 {
+        None
+    } else {
+        let u8path: Vec<u8> = path_buf
+            .iter()
+            .take_while(|c| **c != 0)
+            .map(|c| *c as u8)
+            .collect();
+
+        match str::from_utf8(&u8path) {
+            Ok(p) if p.starts_with("../devices/") => {
+                // readlink returns "../devices/pseudo/..." but the contract
+                // filesystem only wants to know about the "/pseudo/..." part,
+                // so we strip off the first path element before returning it.
+                // We also need to drop any bytes after the terminating NULL.
+                let i8path = path_buf[10..u8path.len() + 1].to_vec();
+                Some((p.to_string(), i8path))
+            }
+            _ => None,
+        }
+    }
+}
+
+impl Template {
+    pub fn new(typ: ContractType) -> Result<Self, crate::ExecutionError> {
+        let path = path(typ, None, "template");
```

Comment:
If modified as above, this could be, `let path = typ.path_prefix().join("template");`